state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -1269,87 +1359,108 @@ pub mod pallet {
-					.expect("CCM that scheduled Swaps must exist in storage");
-				match swap_leg {
-					CcmSwapLeg::Principal => ccm_output.principal = Some(swap_output),
-					CcmSwapLeg::Gas => ccm_output.gas = Some(swap_output),
-				}
-				if let Some((principal, gas)) = ccm_output.completed_result() {
-					Self::schedule_ccm_egress(
-						ccm_id,
-						PendingCcms::<T>::take(ccm_id).expect("Ccm can only be completed once."),
-						(principal, gas),
-					);
-					*maybe_ccm_output = None;
-				}
+		fn schedule_swap(
+			input_asset: Asset,
+			output_asset: Asset,
+			input_amount: AssetAmount,
+			refund_params: Option<SwapRefundParameters>,
+			swap_type: SwapType,
+			swap_request_id: SwapRequestId,
+		) -> SwapId {
+			let swap_id = SwapIdCounter::<T>::mutate(|id| {
+				id.saturating_accrue(1);
+				*id
 			});
-		}
 
-		/// Schedule the egress of a completed Cross chain message.
-		fn schedule_ccm_egress(
-			ccm_id: u64,
-			ccm_swap: CcmSwap,
-			(ccm_output_principal, ccm_output_gas): (AssetAmount, AssetAmount),
-		) {
-			// Schedule the given ccm to be egressed and deposit a event.
-			if let Ok(ScheduledEgressDetails { egress_id, egress_amount, fee_withheld }) =
-				T::EgressHandler::schedule_egress(
-					ccm_swap.destination_asset,
-					ccm_output_principal,
-					ccm_swap.destination_address.clone(),
-					Some((ccm_swap.deposit_metadata, ccm_output_gas)),
-				) {
-				if let Some(swap_id) = ccm_swap.principal_swap_id {
-					Self::deposit_event(Event::<T>::SwapEgressScheduled {
-						swap_id,
-						egress_id,
-						asset: ccm_swap.destination_asset,
-						amount: egress_amount,
-						fee: fee_withheld,
-					});
-				}
-				Self::deposit_event(Event::<T>::CcmEgressScheduled { ccm_id, egress_id });
-			} else {
-				log_or_panic!("CCM egress scheduling should never fail.");
-			}
+			let execute_at = frame_system::Pallet::<T>::block_number() + SWAP_DELAY_BLOCKS.into();
+
+			SwapQueue::<T>::append(
+				execute_at,
+				Swap::new(
+					swap_id,
+					swap_request_id,
+					input_asset,
+					output_asset,
+					input_amount,
+					refund_params,
+					swap_type,
+				),
+			);
+
+			Self::deposit_event(Event::<T>::SwapScheduled {
+				swap_request_id,
+				swap_id,
+				input_amount,
+				swap_type,
+				execute_at,
+			});
+
+			swap_id
 		}
 	}
 
-	impl<T: Config> SwapDepositHandler for Pallet<T> {
+	impl<T: Config> SwapRequestHandler for Pallet<T> {
 		type AccountId = T::AccountId;
 
-		/// Callback function to kick off the swapping process after a successful deposit.
-		fn schedule_swap_from_channel(
-			deposit_address: ForeignChainAddress,
-			deposit_block_height: u64,
-			from: Asset,
-			to: Asset,
-			amount: AssetAmount,
-			destination_address: ForeignChainAddress,
-			broker_commission: Beneficiaries<Self::AccountId>,
+		fn init_swap_request(
+			input_asset: Asset,
+			input_amount: AssetAmount,
+			output_asset: Asset,
+			request_type: SwapRequestType,
+			broker_fees: Beneficiaries<Self::AccountId>,
```

Comment:
Broker fees only apply to swaps from channels, we know that they will never apply to other swap types - why not embed them in the SwapRequestType?

Reply:
We can also have swaps initiated from the contract, where broker fees don't apply. Otherwise would be a clear improvement, but I'm not sure now.